docs/structured-streaming-state-data-source.md

```diff
@@ -0,0 +1,248 @@
+
+The following options must be set for the source.
+
+<table class="table table-striped">
+<thead><tr><th>Option</th><th>value</th><th>meaning</th></tr></thead>
+<tr>
+  <td>path</td>
+  <td>string</td>
+  <td>Specify the root directory of the checkpoint location. You can either specify the path via option("path", `path`) or load(`path`).</td>
+</tr>
+</table>
+
+The following configurations are optional:
+
+<table class="table table-striped">
+<thead><tr><th>Option</th><th>value</th><th>default</th><th>meaning</th></tr></thead>
+<tr>
+  <td>batchId</td>
+  <td>numeric value</td>
+  <td>latest committed batch</td>
+  <td>Represents the target batch to read from. This option is used when users want to perform time-travel. The batch should be committed but not yet cleaned up.</td>
+</tr>
+<tr>
+  <td>operatorId</td>
+  <td>numeric value</td>
+  <td>0</td>
+  <td>Represents the target operator to read from. This option is used when the query is using multiple stateful operators.</td>
+</tr>
+<tr>
+  <td>storeName</td>
+  <td>string</td>
+  <td>DEFAULT</td>
+  <td>Represents the target state store name to read from. This option is used when the stateful operator uses multiple state store instances. It is not required except stream-stream join.</td>
+</tr>
+<tr>
+  <td>joinSide</td>
+  <td>string ("left" or "right")</td>
+  <td>(none)</td>
+  <td>Represents the target side to read from. This option is used when users want to read the state from stream-stream join.</td>
+</tr>
+</table>
+
+### Reading state for Stream-stream join
+
+Structured Streaming implements the feature stream-stream join via leveraging multiple instances of state store internally.
+These instances logically compose buffers to store the input rows for left and right.
+
+Since it is more obvious to users to reason about, the data source provides the option 'joinSide' to read the buffered input for specific side of the join.
+To enable the functionality to read the internal state store instance, we also allow specifying the option 'storeName', with restriction that 'storeName' and 'joinSide' cannot be specified together.
+
+## State metadata source
+
+Before querying the state from existing checkpoint via state data source, users would like to understand the information for the checkpoint, especially about state operator. This includes which operators and state store instances are available in the checkpoint, available range of batch IDs, etc.
+
+Structured Streaming provides a data source named "State metadata source" to provide the state-related metadata information from the checkpoint.
+
+Note: The metadata is constructed when the streaming query is running with Spark 4.0+. The existing checkpoint which has been running with lower Spark version does not have the metadata and will be unable to query/use with this metadata source. It is required to run the streaming query pointing the existing checkpoint in Spark 4.0+ to construct the metadata before querying.
```

Comment:
Does the last sentence mean, for a checkpoint running with lower Spark versions, once we run the streaming query with the checkpoint using Spark 4.0+, the metadata will be constructed so we can use "state metadata source" on it after that?

Reply:
Yes, Spark 4.0+ will build a state metadata if there is no state metadata available for the stateful operator in the checkpoint, hence checkpoint being created with older version of Spark will get the state metadata once it is run with Spark 4.0+.